lib/pbench/test/unit/server/auth/test_auth.py

```diff
@@ -34,3 +44,58 @@ def test_token_introspect_wrong_aud_claim(self, keycloak_oidc, keycloak_mock_tok
                 options=options,
             )
         assert str(e.value) == "Invalid audience"
+
+    def test_token_introspect_invalid_algorithm(
+        self, keycloak_oidc, keycloak_mock_token
+    ):
+        options = {"verify_signature": True, "verify_aud": True, "verify_exp": True}
+        with pytest.raises(InvalidAlgorithmError) as e:
+            keycloak_oidc.token_introspect_offline(
+                token=keycloak_mock_token,
+                key="some_secret",
+                algorithms=["INVALID_ALGORITHM"],
+                audience="wrong_client",
+                options=options,
+            )
+        assert str(e.value) == "The specified alg value is not allowed"
+
+    def test_token_introspect_expired_signature(self, keycloak_oidc):
+        current_utc = datetime.datetime.now(datetime.timezone.utc)
+        payload = {
+            "iat": current_utc,
+            "exp": current_utc - datetime.timedelta(minutes=1),  # expired token
+            "sub": "12345",
+            "aud": "test_client",
+        }
+
+        # Get jwt key
+        token = jwt.encode(payload, key="some_secret", algorithm="HS256")
+        options = {"verify_signature": True, "verify_aud": True, "verify_exp": True}
+        with pytest.raises(ExpiredSignatureError) as e:
+            keycloak_oidc.token_introspect_offline(
+                token=token,
+                key="some_secret",
+                algorithms=["HS256"],
+                audience="wrong_client",
+                options=options,
+            )
+        assert str(e.value) == "Signature has expired"
+
+    def test_token_introspect_invalid_token(self, monkeypatch, keycloak_oidc):
+        def offline_token_introspection_exception(
+            self, token: str, key: str, audience: str
+        ):
+            raise InvalidTokenError("Invalid token")
+
+        monkeypatch.setattr(
+            OpenIDClient,
+            "token_introspect_offline",
+            offline_token_introspection_exception,
+        )
+        with pytest.raises(InvalidTokenError) as e:
+            keycloak_oidc.token_introspect_offline(
+                token="token",
+                key="some_secret",
+                audience="wrong_client",
+            )
+        assert str(e.value) == "Invalid token"
```

Comment:
These are probably great tests to have (I haven't really looked into them...), but what I was referring to in [my comment](https://github.com/distributed-system-analysis/pbench/pull/3056#issuecomment-1373904963) was unit testing for the `verify_third_party_token()` code which _calls_ `token_introspect_online()` and `token_introspect_offline()`.  (My apologies for being unclear; did you click on the hyperlink and look at the coverage details?)

Reply:
Yes, but these 3 were something I just thought would be good to have when I started looking into the unit tests for the auth so I added them. The one that might actually change the code coverage would be I think `test_online_third_party_verification`